keras_nlp/samplers/contrastive_sampler.py

```diff
@@ -0,0 +1,232 @@
+
+            # Replicate for `self.k` times to find the best token in top-k
+            # candidates.
+            prompt_beams = create_beams(prompt)
+            mask_beams = create_beams(mask)
+            hidden_states_beams = create_beams(hidden_states)
+            cache_beams = tf.nest.map_structure(create_beams, cache)
+
+            # Get top-k candidate tokens and their probabilities.
+            top_k_probabilities, top_k_indices = tf.math.top_k(
+                probabilities, k=self.k, sorted=False
+            )
+            next_token_probabilities = flatten_beams(top_k_probabilities)
+            next_token = flatten_beams(top_k_indices)
+            next_token = tf.cast(next_token, prompt.dtype)
+            next_token = tf.where(
+                mask_beams[:, index], prompt_beams[:, index], next_token
+            )
+
+            # Update the prompt with the next token.
+            next_token = next_token[:, tf.newaxis]
+            prompt_beams = dynamic_update_slice(
+                prompt_beams, next_token, [0, index]
+            )
+
+            # Compute the logits and hidden states for top-k candidate tokens.
+            next_logits, next_hidden_states_beams, cache_beams = next(
+                prompt_beams, cache_beams, index + 1
+            )
+
+            # Compute the max similarity score for top-k candidate tokens
+            # against previous tokens.
+            last_token_state = next_hidden_states_beams
+            previous_states = hidden_states_beams[:, :index, :]
+            similarity_scores = self.similarity(
+                previous_states, last_token_state
+            )
+            max_similarity_scores = tf.cast(
+                tf.reduce_max(similarity_scores, axis=1),
+                dtype=next_token_probabilities.dtype,
+            )
+            if index == 0:
+                # If the index is 0, there is no previous states so we set
+                # `max_similarity_scores` the same for all beams.
+                max_similarity_scores = tf.zeros_like(max_similarity_scores)
+            # The final score of each candidate token is weighted sum of
+            # probability and similarity against previous tokens.
+            accumulated_scores = (
+                (1 - self.alpha) * next_token_probabilities
+                - self.alpha * max_similarity_scores
+            )
+            # Unflatten varibles to shape [batch_size, self.k, ...] for
+            # gather purpose.
+            unflat_score = unflatten_beams(accumulated_scores)
+            unflat_prompt = unflatten_beams(prompt_beams)
+            unflat_next_logits = unflatten_beams(next_logits)
+            unflat_next_hidden_states = unflatten_beams(
+                next_hidden_states_beams
+            )
+            unflat_cache = tf.nest.map_structure(unflatten_beams, cache_beams)
+            win_token_indices = tf.math.argmax(unflat_score, axis=1)
```

Comment:
win feels a bit odd here. `best_token_indices` (is that what this variable is?)

Reply:
yea, I find it hard to name it, `best_token_indices` is better, I am going for it.